api/api-doc/storage_service.json

```diff
@@ -802,7 +802,7 @@
          "operations":[
             {
                "method":"POST",
-               "summary":"Trigger a global cluster cleanup through the topology coordinator",
+               "summary":"Trigger a global cleanup",
```

Comment:
The way it is now with this change that this api is cluster-wide with topology_coordinator_enabled or node-local would be extremely confusing to the poor user.

And I really dislike that `cleanup_all` is doing cluster-wide cleanup as it's not apparent in its name.

I think it would be best to rename and separate the apis into:
- `cluster_cleanup` for the former case, that would fail if `!ss.is_topology_coordinator_enabled()`, and
- `cleanup` to cleanup all non-local keyspaces on the local node, similar to `compact` vs. `keyspace_compaction` for major compaction and to `flush` vs. `keyspace_flush`.

@kostja @gleb-cloudius @kbr-scylla @tgrabiec @denesb @vladzcloudius @amnonh WDYT?

(for the record, I don't think this is bike-shedding but rather a UX issue we should resolve before raft-topology-changes go out of experimental mode)

Reply:
Otherwise, LGTM

Reply:
I'm for the renames

Reply:
i concur with this suggestion. as the existing `cleanup_all` asks all nodes in the ring to perform global cleanup. while the fall back performs global cleanup locally.

but i'd wait for Gleb and other developers inputs before reverting the change, i.e., to add the `cleanup` API back,  and adding another commit to rename the `cleanup_all` API to `cluster_cleanup`.

also, shall we expose `cluster_cleanup` with the existing `nodetool cleanup` command? as its semantic is different from the local global cleanup. probably we can have an option for exposing it, something like `nodetool cleanup --cluster` ? and do the local global cleanup with `nodetool cleanup`.

Reply:
Unless we introduce new nodetool command for global cleanup that works only with raft I do not see the point to have to different APIs. Currently `clenaup_all` is called when "nodetool cleanup" is executed without any parameters and user are used to do on each node. With raft it is unnecessary, but not a big deal because only the first call will do the cleanup. All others will either do nothing or even fail with an error that cleanup is already running (actually we need to improve the error the user will currently get). 

So unless we want to re-educate users to run cleanup on one node only in case of raft topology having to different APIs is not better.

Reply:
People can use the API:s directly, not just through nodetool. So they should have non-surprising behavior, we cannot justify confusion by "users using nodetool cleanup will not notice any changes".

Reply:
> People can use the API:s directly, not just through nodetool. So they should have non-surprising behavior, we cannot justify confusion by "users using nodetool cleanup will not notice any changes".

We have three options: 
1. have two different API points and use different one in "nodetool cleanup" depending on raft topology setting
2. have on API and run different code depending on raft topology setting
3. have two different APIs and run non raft api in "nodetool cleanup" and add different nodetool verb to run raft one.

All of them are bad. All of then will confuse someone. Option 1 will not be any less confusing for nodetool users and those are prevailing. Option 3 will make raft API adoption minimal despite it been the most performant option when raft is enable.

Reply:
Indeed, option 2 (current state, as of this patch) seems to be the least bad. We don't want to go change nodetool, not until we get rid of the java one. Coordinating a change across 3 repos is a pain.

Reply:
nodetool, as its name suggests, is a per-node tool.
I'd expect an entity like scylla-manager to be aware of the cluster_cleanup api and use it if it's available, otherwise the semantics should not change for the existing users that would call `nodetool cleanup` on all nodes.

per-node cleanup should be safe to run in parallel on all nodes, and if the cluster-cleanup is enabled other nodes would return an error.  If you force the user to run it sequentially, first it can take incredibly long time 
I wish it would do nothing on other nodes if cluster cleanup is already done, but our detection algorithm for which sstable requires cleanup is heuristic and has a lot of false positive so you'd end up with sstables being cleaned up over and over again.

Reply:
> Indeed, option 2 (current state, as of this patch) seems to be the least bad. We don't want to go change nodetool, not until we get rid of the java one. Coordinating a change across 3 repos is a pain.

java-based nodetool+scylla-jmx invoke keyspace_cleanup so there's no issue there and nothing needs to change.
The point here is about `scylla-nodetool cleanup` calling the global cleanup api.
I see no problem with presenting `scylla-nodetool cluster-cleanup`, that may fail if scylla doesn't support it, and `scylla-nodetool cleanup` that calls the node-local cleanup api that cleans up all keyspaces.

Reply:
> nodetool, as its name suggests, is a per-node tool.

We do not have cluster tool, so nodetool is responsible to initiate clusterwide operations as well. For instance removenode does not in fact remove the node it was issued against, but some other node.

> I'd expect an entity like scylla-manager to be aware of the cluster_cleanup api and use it if it's available, otherwise the semantics should not change for the existing users that would call `nodetool cleanup` on all nodes.

Unless scylla-manager is part of Scylla packaging it is absolutely irrelevant.

> 
> per-node cleanup should be safe to run in parallel on all nodes, and if the cluster-cleanup is enabled other nodes would return an error.

No they will not. Read this: https://github.com/scylladb/scylladb/pull/16757#discussion_r1466017492

> If you force the user to run it sequentially, first it can take incredibly long time I wish it would do nothing on other nodes if cluster cleanup is already done, but our detection algorithm for which sstable requires cleanup is heuristic and has a lot of false positive so you'd end up with sstables being cleaned up over and over again.

Because those algorithms are heuristic it is advisable to use global cleanup mechanism when available otherwise some cleanups will run needlessly.

Reply:
> Because those algorithms are heuristic it is advisable to use global cleanup mechanism when available otherwise some cleanups will run needlessly.

I agree.  It is indeed the best way to run it, but if one would run it on each node in its turn, I think it's going to do a lot of needless work when rerunning after a successful cluster-wide cleanup_all.

Reply:
> but if one would run it on each node in its turn

What is "it" in this sentence?

Reply:
> > but if one would run it on each node in its turn
> 
> What is "it" in this sentence?

nodetool cleanup (that invokes the cleanup_all api)

Reply:
> > > but if one would run it on each node in its turn
> > 
> > 
> > What is "it" in this sentence?
> 
> nodetool cleanup (that invokes the cleanup_all api)

With raft enabled "cleanup_all" will do anything at all only the first time it is invoked (assuming cleanup is needed). After that it will do nothing until a topology operation that leaves some nodes dirty again.

Reply:
i think the crux of the problem is that the users won't have a clear expectation for the output of the `cleanup` subcommand. will it clear the local node only or will it clear all nodes? unless they are aware which API is called under the hood, they have no idea what should the result.